libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -170,38 +213,58 @@ internal object MapboxNavigationTelemetry {
 
         override fun onFinalDestinationArrival(routeProgress: RouteProgress) {
             log("onFinalDestinationArrival")
-            this@MapboxNavigationTelemetry.routeProgress = routeProgress
+            this@MapboxNavigationTelemetry.routeData.routeProgress = routeProgress
             processArrival()
         }
     }
 
     private val navigationSessionStateObserver = NavigationSessionStateObserver { sessionState ->
         log("session state is $sessionState")
+        val legacyState = this.sessionState
+        this.sessionState = sessionState
         when (sessionState) {
-            IDLE, FREE_DRIVE -> {
-                sessionStop()
-                handleStateChanged(this.sessionState, sessionState)
+            IDLE,
+            FREE_DRIVE -> {
+                if (legacyState == ACTIVE_GUIDANCE) {
+                    handleCancelNavigation()
+                }
+                resetLocalVariables()
+                resetDynamicValues()
             }
-            ACTIVE_GUIDANCE -> {
-                locationsCollector.flushBuffers()
-                handleStateChanged(this.sessionState, sessionState)
-                needStartSession = true
-                startSessionIfNeedAndCan()
+            ACTIVE_GUIDANCE -> Unit // do nothing
+        }
+
+        when (val freeDriveEvent = getFreeDriveEvent(legacyState, this.sessionState)) {
+            START -> {
+                handleTelemetryState()
+                trackFreeDrive(freeDriveEvent)
+            }
+            STOP -> {
+                trackFreeDrive(freeDriveEvent)
+                handleTelemetryState()
```

Comment:
Why is the method call order swapped between these two?

Reply:
This is a bit ticky: 
- if freeDrive is started -> need apply all ids/params(sessionIdentifier, modeId and so on) before track START of freeDrive;
- if freeDrive is stopped -> need track freeDrive before new params are applied(like sessionIdentifier, modeId and so on).